spark/src/main/scala/org/apache/spark/sql/delta/UpdateExpressionsSupport.scala

```diff
@@ -387,4 +407,110 @@ trait UpdateExpressionsSupport extends CastSupport with SQLConfHelper with Analy
         }
     }
   }
+
+  /**
+   * Replaces 'CastSupport.cast'. Selects a cast based on 'spark.sql.storeAssignmentPolicy' if
+   * 'spark.databricks.delta.updateAndMergeCastingFollowsAnsiEnabledFlag. is false, and based on
+   * 'spark.sql.ansi.enabled' otherwise.
+   */
+  private def cast(child: Expression, dataType: DataType, columnName: String): Expression = {
+    if (conf.getConf(DeltaSQLConf.UPDATE_AND_MERGE_CASTING_FOLLOWS_ANSI_ENABLED_FLAG)) {
+      return Cast(child, dataType, Option(conf.sessionLocalTimeZone))
+    }
+
+    conf.storeAssignmentPolicy match {
+      case SQLConf.StoreAssignmentPolicy.LEGACY =>
+        Cast(child, dataType, Some(conf.sessionLocalTimeZone), ansiEnabled = false)
+      case SQLConf.StoreAssignmentPolicy.ANSI =>
+        val cast = Cast(child, dataType, Some(conf.sessionLocalTimeZone), ansiEnabled = true)
+        if (canCauseCastOverflow(cast)) {
+          CheckOverflowInTableWrite(cast, columnName)
+        } else {
+          cast
+        }
+      case SQLConf.StoreAssignmentPolicy.STRICT =>
+        UpCast(child, dataType)
+    }
+  }
+
+  private def containsIntegralOrDecimalType(dt: DataType): Boolean = dt match {
+    case _: IntegralType | _: DecimalType => true
+    case a: ArrayType => containsIntegralOrDecimalType(a.elementType)
+    case m: MapType =>
+      containsIntegralOrDecimalType(m.keyType) || containsIntegralOrDecimalType(m.valueType)
+    case s: StructType =>
+      s.fields.exists(sf => containsIntegralOrDecimalType(sf.dataType))
+    case _ => false
+  }
+
+  private def canCauseCastOverflow(cast: Cast): Boolean = {
+    containsIntegralOrDecimalType(cast.dataType) &&
+      !Cast.canUpCast(cast.child.dataType, cast.dataType)
+  }
+}
+
+case class CheckOverflowInTableWrite(child: Expression, columnName: String)
```

Comment:
Seems to be almost the same as `CheckOverflowInTableInsert` in Spark, could we inherit it?

Reply:
I was indeed looking into it, but there are some problems with that
* This case is a case class and CheckOverflowInTableInsert is one as well, so Scala does not like that
* I cannot refactor CheckOverflowInTableInsert because it is part of the OSS Spark imports. This also inhibits me from making methods like getCast protected or the error throwing function configurable
* Even if I could overwrite methods, with doGenCodeWithBetterErrorMsg and eval I would have to overwrite the main ones so most of the code would not be shared (this is more of a minor thing though)